File: backend-fastapi/lib/pydantic_/filters.py
from datetime import datetime
from typing import (
    Annotated,
    Any,
    Callable,
    Generic,
    Optional,
    TypeVar,
    cast,
    get_args,
)

from pydantic import (
    AfterValidator,
    BeforeValidator,
    EmailStr,
    Field,
    TypeAdapter,
)
from pydantic.fields import FieldInfo
from pydantic_core import PydanticCustomError

from lib.types_ import Filter, FilterOperation
from lib.utils import check_bool

T = TypeVar("T")

check_filter_op: Callable = TypeAdapter(FilterOperation).validate_python


def _numeric_filter_validator(
    op: FilterOperation, raw: str, adapter: TypeAdapter, is_index: bool = False
) -> Filter:
    if is_index:
        return _index_filter_validator(op, raw, adapter)

    if op in ["eq", "ne", "gt", "gte", "lt", "lte"]:
        val = adapter.validate_python(raw)
        return Filter(op=op, val=val)
    elif op in ["in", "nin"]:
        l = raw.split(",")
        val = [adapter.validate_python(item) for item in l]
        return Filter(op=op, val=val)
    elif op == "null":
        val = check_bool(raw)
        return Filter(op=op, val=val)
    else:
        raise PydanticCustomError(
            "invalid numeric operation",
            f"{op} is not a valid operation for numeric fields - Valid: eq,ne,gt,gte,lt,lte,in,nin,null",
        )


def _index_filter_validator(
    op: FilterOperation, raw: str, adapter: TypeAdapter
) -> Filter:
    if op in ["eq", "ne"]:
        val = adapter.validate_python(raw)
        return Filter(op=op, val=val)
    elif op in ["in", "nin"]:
        l = raw.split(",")
        val = [adapter.validate_python(item) for item in l]
        return Filter(op=op, val=val)
    elif op == "null":
        val = check_bool(raw)
        return Filter(op=op, val=val)
    else:
        raise PydanticCustomError(
            "invalid ObjectId operation",
            f"{op} is not a valid operation for linked fields - Valid: eq,ne,null,in,nin",
        )


def _string_filter_validator(
    op: FilterOperation,
    raw: str,
    adapter: TypeAdapter,
) -> Filter:
    if op in ["eq", "ne"]:
        val = adapter.validate_python(raw)
        return Filter(op=op, val=val)
    elif op in ["in", "nin"]:
        l = raw.split(",")
        val = [adapter.validate_python(item) for item in l]
        return Filter(op=op, val=val)
    elif op == "null":
        val = check_bool(raw)
        return Filter(op=op, val=val)
    elif op in ["like", "ilike"]:
        return Filter(op=op, val=raw)
    else:
        raise PydanticCustomError(
            "invalid string operation",
            f"{op} is not a valid operation for string fields - Valid: eq,ne,in,nin,null,like,ilike",
        )


def _boolean_filter_validator(
    op: FilterOperation, raw: str, adapter: TypeAdapter
) -> Filter:
    if op in ["eq", "ne", "null"]:
        val = check_bool(raw)
        return Filter(op=op, val=val)
    else:
        raise PydanticCustomError(
            "invalid boolean operation",
            f"{op} is not a valid operation for boolean fields - Valid: eq,ne,null",
        )


def _datetime_filter_validator(
    op: FilterOperation, raw: str, adapter: TypeAdapter
) -> Filter:
    if op in ["eq", "ne", "gt", "gte", "lt", "lte"]:
        val = adapter.validate_python(raw)
        return Filter(op=op, val=val)
    elif op in ["in", "nin"]:
        l = raw.split(",")
        val = [adapter.validate_python(item) for item in l]
        return Filter(op=op, val=val)
    elif op == "null":
        val = check_bool(raw)
        return Filter(op=op, val=val)
    else:
        raise PydanticCustomError(
            "invalid datetime operation",
            f"{op} is not a valid operation for datetime fields - Valid: eq,ne,gt,gte,lt,lte,in,nin,null",
        )


def _get_field_info(field) -> FieldInfo | None:
    extra = get_args(field)[1:]
    for metadata in extra:
        if isinstance(metadata, FieldInfo):
            return metadata
    return None


def _extract_raw_filter(value: str) -> tuple[FilterOperation, Any]:
    if ":" in value:
        op, raw_val = value.split(":", 1)
    else:
        op, raw_val = "eq", value

    check_filter_op(op)
    return cast(FilterOperation, op), raw_val


def _make_filter_validator(real_type: Any):
    base_class = get_args(real_type)[0]
    annotations = _get_field_info(real_type)
    json_schema_extra = getattr(annotations, "json_schema_extra") or {}
    is_index: bool = json_schema_extra.get("is_index", False)
    adapter = TypeAdapter(real_type)

    if base_class not in [int, float, str, EmailStr, bool, datetime]:
        raise RuntimeError(f"Unknow base type {base_class}")

    def validator(value: str | dict) -> Filter:
        if isinstance(value, str):
            op, raw_val = _extract_raw_filter(value)
        else:
            try:
                op = value["op"]
                raw_val = value["val"]
            except KeyError:
                message: str = (
                    f"Field Filter {value} is not valid. It must define the op and val fields"
                )
                raise PydanticCustomError("Invalid Field Filter", message)  # type: ignore

        if base_class in [int, float]:
            return _numeric_filter_validator(op, raw_val, adapter, is_index)
        elif base_class in [str, EmailStr]:
            return _string_filter_validator(op, raw_val, adapter)
        elif base_class in [bool]:
            return _boolean_filter_validator(op, raw_val, adapter)
        elif base_class in [datetime]:
            return _datetime_filter_validator(op, raw_val, adapter)
        else:
            # This should not be reached. Linting purposes
            raise RuntimeError(f"Unknow base type {base_class}")

    return validator


class HttpFilter(Generic[T]):
    def __class_getitem__(cls, item):
        field_info = _get_field_info(item)
        extra = getattr(field_info, "json_schema_extra", None) or {}
        examples = extra.get("filter_examples", None)
        if not examples and field_info is not None:
            examples = field_info.examples

        return Annotated[
            str | Filter,
            BeforeValidator(_make_filter_validator(item)),
            Field(examples=examples),
        ]


def _field_filters_validator(filters: list[Filter]):
    used_operators: list[FilterOperation] = []
    errors: list[str] = []
    duplicate: list[str] = []

    # Rule 0: make sure no operator is used twice
    for filter in filters:
        op = filter["op"]
        if op in used_operators and op not in duplicate:
            duplicate.append(op)
            errors.append(
                f"cannot use an operator twice for the same field. {op} used multiple times"
            )
        used_operators.append(op)

    length = len(used_operators)
    eq_used = "eq" in used_operators

    # Rule 1: eq should be used exclusively
    if length >= 2 and eq_used:
        errors.append(
            f"eq can only be used exclusively. {used_operators} used at the same time"
        )

    # Rule 2: if eq not used than null should be used exclusively
    elif not eq_used and "null" in used_operators and length >= 2:
        errors.append(
            f"null operator should be used exclusively. {used_operators} used at the same time"
        )

    # Rule 3: if eq not used than in should be used exclusively
    elif not eq_used and "in" in used_operators and length >= 2:
        errors.append(
            f"in operator should be used exclusively. {used_operators} used at the same time"
        )

    # Rule 4: gt/gte cannot be used together
    if "gt" in used_operators and "gte" in used_operators:
        errors.append("gt and gte operators should not be used together")

    # Rule 5: lt/lte cannot be used together
    if "lt" in used_operators and "lte" in used_operators:
        errors.append("lt and lte operators should not be used together")

    # Rule 6: like/ilike cannot be used together
    if "like" in used_operators and "ilike" in used_operators:
        errors.append("like and ilike operators should not be used together")

    return filters


class HttpFilters(Generic[T]):
    def __class_getitem__(cls, item):
        field_info = _get_field_info(item)
        description = getattr(field_info, "description", "")
        extra = getattr(field_info, "json_schema_extra", None) or {}
        examples = extra.get("filter_examples", None)
        if examples is None:
            field_info = _get_field_info(item)
            if field_info is not None:
                examples = field_info.examples
            else:
                examples = []

        return Annotated[
            Optional[list[HttpFilter[item]]],
            Field(
                None,
                description=description,
                json_schema_extra=extra,
                examples=examples,
            ),
            AfterValidator(_field_filters_validator),
        ]

File: backend-express/src/lib/zod_/filters.ts
import { RefinementCtx, ZodNever, ZodTypeAny } from "zod";

import { Filter, FilterOperation } from "../types";
import { zod } from "./base";

type QueryParamTypes = "numeric" | "string" | "boolean" | "date";

type TransformOption = { isIndex?: boolean };

type OpenapiDoc = { description?: string; example?: any };

const updateContextFromError = (
    ctx: RefinementCtx,
    err: any,
    message: string
): ZodNever => {
    let msg: any;
    try {
        msg = JSON.parse(err.message);
    } catch {
        msg = message;
    }

    ctx.addIssue({ code: zod.ZodIssueCode.custom, message: msg });
    return zod.NEVER;
};

const numericQueryParamTransform = (
    field: ZodTypeAny,
    value: any,
    context: RefinementCtx,
    options?: TransformOption
): { op: FilterOperation; val: number | number[] | boolean } | ZodNever => {
    if (options?.isIndex) {
        return indexQueryParamTransform(field, value, context);
    }

    if (!value.includes(":")) {
        try {
            return { op: "eq", val: field.parse(Number(value)) };
        } catch (err) {
            return updateContextFromError(
                context,
                err,
                "Invalid number format"
            );
        }
    }
    const [op, val] = value.split(":");

    switch (op) {
        case "eq":
        case "ne":
        case "gt":
        case "gte":
        case "lt":
        case "lte":
            try {
                return { op, val: field.parse(Number(val)) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid number for ${op} operator`
                );
            }

        case "in":
        case "nin":
            try {
                const vals = val.split(",");
                return { op, val: zod.array(field).parse(vals) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid numeric options for ${op} operator`
                );
            }

        case "null":
            try {
                const boolValue = zod.coerce.boolean().parse(val);
                field.parse(boolValue);
                return { op, val: boolValue };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid boolean: ${val}`
                );
            }

        default:
            context.addIssue({
                code: zod.ZodIssueCode.custom,
                message: `Unknown operator "${op}". Valid: eq,ne,gt,gte,lt,lte,in,nin,null`,
            });
            return zod.NEVER;
    }
};

const indexQueryParamTransform = (
    field: ZodTypeAny,
    value: any,
    context: RefinementCtx
): { op: FilterOperation; val: number | number[] | boolean } | ZodNever => {
    if (!value.includes(":")) {
        try {
            return { op: "eq", val: field.parse(Number(value)) };
        } catch (err) {
            return updateContextFromError(
                context,
                err,
                "Invalid index format - should be a number"
            );
        }
    }
    const [op, val] = value.split(":");

    switch (op) {
        case "eq":
        case "ne":
            try {
                return { op, val: field.parse(Number(val)) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid index for ${op} operator`
                );
            }

        case "in":
        case "nin":
            try {
                const vals = val.split(",");
                return { op, val: zod.array(field).parse(vals) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid index options for ${op} operator`
                );
            }

        case "null":
            try {
                const boolValue = zod.coerce.boolean().parse(val);
                field.parse(boolValue);
                return { op, val: boolValue };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid boolean: ${val}`
                );
            }

        default:
            context.addIssue({
                code: zod.ZodIssueCode.custom,
                message: `Unknown operator "${op}". Valid: eq,ne,null,in,nin`,
            });
            return zod.NEVER;
    }
};

const stringQueryParamTransform = (
    field: ZodTypeAny,
    value: any,
    context: RefinementCtx
): { op: FilterOperation; val: string | string[] | boolean } | ZodNever => {
    if (!value.includes(":")) {
        try {
            return { op: "eq", val: field.parse(value) };
        } catch (err) {
            return updateContextFromError(
                context,
                err,
                "Invalid string format"
            );
        }
    }
    const [op, val] = value.split(":");

    switch (op) {
        case "eq":
        case "ne":
            try {
                return { op, val: field.parse(val) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid string for ${op} operator`
                );
            }

        case "in":
        case "nin":
            try {
                const vals = val.split(",");
                return { op, val: zod.array(field).parse(vals) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid string options for ${op} operator`
                );
            }

        case "null":
            try {
                const boolValue = zod.coerce.boolean().parse(val);
                field.parse(boolValue);
                return { op, val: boolValue };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid boolean: ${val}`
                );
            }

        case "like":
        case "ilike":
            return { op, val };

        case "regex":
            try {
                new RegExp(val);
                return { op, val };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid regex pattern: ${val}`
                );
            }

        default:
            context.addIssue({
                code: zod.ZodIssueCode.custom,
                message: `Unknown operator "${op}". Valid: eq,ne,in,nin,null,like,ilike,regex`,
            });
            return zod.NEVER;
    }
};

const booleanQueryParamTransform = (
    field: ZodTypeAny,
    value: any,
    context: RefinementCtx
): { op: FilterOperation; val: boolean } | ZodNever => {
    if (!value.includes(":")) {
        try {
            return { op: "eq", val: field.parse(value === "true") };
        } catch (err) {
            return updateContextFromError(
                context,
                err,
                "Invalid boolean format (use 'true' or 'false')"
            );
        }
    }
    const [op, val] = value.split(":");

    switch (op) {
        case "eq":
        case "ne":
        case "null":
            try {
                const boolValue = zod.coerce.boolean().parse(val);
                field.parse(boolValue);
                return { op, val: boolValue };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid boolean: ${val}`
                );
            }

        default:
            context.addIssue({
                code: zod.ZodIssueCode.custom,
                message: `Unknown operator "${op}". Valid: eq,ne,null`,
            });
            return zod.NEVER;
    }
};

const dateQueryParamTransform = (
    field: ZodTypeAny,
    value: any,
    context: RefinementCtx
): { op: FilterOperation; val: Date | Date[] | boolean } | ZodNever => {
    if (!value.includes(":")) {
        try {
            const date = new Date(value);
            return { op: "eq", val: field.parse(date) };
        } catch (err) {
            return updateContextFromError(
                context,
                err,
                "Invalid date format (use ISO 8601)"
            );
        }
    }
    const [op, val] = value.split(":");

    switch (op) {
        case "eq":
        case "ne":
        case "gt":
        case "gte":
        case "lt":
        case "lte":
            try {
                const date = new Date(val);
                return { op, val: field.parse(date) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid date for ${op} operator`
                );
            }

        case "in":
        case "nin":
            try {
                const vals = val.split(",");
                return { op, val: zod.array(field).parse(vals) };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid date options format for ${op} operator`
                );
            }

        case "null":
            try {
                const boolValue = zod.coerce.boolean().parse(val);
                field.parse(boolValue);
                return { op, val: boolValue };
            } catch (err) {
                return updateContextFromError(
                    context,
                    err,
                    `Invalid boolean: ${val}`
                );
            }

        default:
            context.addIssue({
                code: zod.ZodIssueCode.custom,
                message: `Unknown operator "${op}". Valid: eq,ne,gt,gte,lt,lte,in,nin,null`,
            });
            return zod.NEVER;
    }
};

const guessType = (field: ZodTypeAny): QueryParamTypes => {
    if (field instanceof zod.ZodNumber) {
        return "numeric";
    } else if (field instanceof zod.ZodString) {
        return "string";
    } else if (field instanceof zod.ZodBoolean) {
        return "boolean";
    } else if (field instanceof zod.ZodDate) {
        return "date";
    } else if (field instanceof zod.ZodTransformer) {
        return guessType(field._def.schema);
    } else if (field instanceof zod.ZodOptional) {
        return guessType(field._def.innerType);
    } else {
        throw new Error(`Could not guess the type of field ${field}`);
    }
};

export const httpFilter = (
    field: ZodTypeAny,
    doc?: OpenapiDoc,
    options?: TransformOption
): ZodTypeAny => {
    let transformFunction: (val: any, ctx: RefinementCtx) => any;
    const baseType = guessType(field);
    if (baseType === "numeric") {
        transformFunction = (val, ctx) =>
            numericQueryParamTransform(field, val, ctx, options);
    } else if (baseType === "string") {
        transformFunction = (val, ctx) =>
            stringQueryParamTransform(field, val, ctx);
    } else if (baseType === "boolean") {
        transformFunction = (val, ctx) =>
            booleanQueryParamTransform(field, val, ctx);
    } else if (baseType === "date") {
        transformFunction = (val, ctx) =>
            dateQueryParamTransform(field, val, ctx);
    } else {
        throw new Error(`Cannot use type ${field} as QueryParam`);
    }
    doc = doc || {};
    let { description, example } = doc;
    description = description || "";
    example = example || null;
    return zod
        .string()
        .openapi({ description, example })
        .transform(transformFunction);
};

const toFieldFilter = (filters: Filter[], ctx: RefinementCtx): Filter[] => {
    const usedOperators: FilterOperation[] = [];
    filters.forEach((f) => {
        // make sure no operator is used twice
        if (usedOperators.includes(f.op)) {
            ctx.addIssue({
                code: zod.ZodIssueCode.custom,
                message: `cannot use an operator twice for the same field. ${f.op} used multiple times`,
            });
        }
        usedOperators.push(f.op);
    });

    const length = usedOperators.length;
    const eqUsed = usedOperators.includes("eq");

    // Rule 1: eq should be used exclusively
    if (length >= 2 && eqUsed) {
        ctx.addIssue({
            code: zod.ZodIssueCode.custom,
            message: `eq can only be used exclusively. ${usedOperators} used at the same time`,
        });
    }

    // Rule 2: if eq not used than null should be used exclusively
    if (!eqUsed && usedOperators.includes("null") && length >= 2) {
        ctx.addIssue({
            code: zod.ZodIssueCode.custom,
            message: `null operator should be used exclusively. ${usedOperators} used at the same time`,
        });
    }

    // Rule 3: if eq not used than in should be used exclusively
    if (!eqUsed && usedOperators.includes("in") && length >= 2) {
        ctx.addIssue({
            code: zod.ZodIssueCode.custom,
            message: `in operator should be used exclusively. ${usedOperators} used at the same time`,
        });
    }

    // Rule 4: gt/gte cannot be used together
    if (usedOperators.includes("gt") && usedOperators.includes("gte")) {
        ctx.addIssue({
            code: zod.ZodIssueCode.custom,
            message: `gt and gte operators should not be used together. ${usedOperators} used at the same time`,
        });
    }

    // Rule 5: lt/lte cannot be used together
    if (usedOperators.includes("lt") && usedOperators.includes("lte")) {
        ctx.addIssue({
            code: zod.ZodIssueCode.custom,
            message: `lt and lte operators should not be used together. ${usedOperators} used at the same time`,
        });
    }

    // Rule 6: like/ilike cannot be used together
    if (usedOperators.includes("like") && usedOperators.includes("ilike")) {
        ctx.addIssue({
            code: zod.ZodIssueCode.custom,
            message: `like and ilike operators should not be used together. ${usedOperators} used at the same time`,
        });
    }

    // Rule 7: regex cannot be used with like/ilike
    if (
        usedOperators.includes("regex") &&
        (usedOperators.includes("ilike") || usedOperators.includes("ilike"))
    ) {
        ctx.addIssue({
            code: zod.ZodIssueCode.custom,
            message: `regex should not be used along like or ilike operators. ${usedOperators} used at the same time`,
        });
    }

    return filters;
};

export const httpFilters = (
    field: ZodTypeAny,
    doc?: OpenapiDoc,
    options?: TransformOption
): ZodTypeAny => {
    return zod.array(httpFilter(field, doc, options)).transform(toFieldFilter);
};
